src/sentry/snuba/metrics/query.py

```diff
@@ -0,0 +1,135 @@
+""" Classes needed to build a metrics query. Inspired by snuba_sdk.query. """
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Flag, auto
+from typing import Literal, Optional, Sequence, Union
+
+from snuba_sdk import Direction, Granularity, Limit, Offset
+from snuba_sdk.conditions import ConditionGroup
+
+from .utils import MetricOperationType
+
+# TODO: Add __all__ to be consistent with sibling modules
+
+
+@dataclass(frozen=True)
+class MetricField:
+    op: Optional[MetricOperationType]
+    metric_name: str
+
+
+class Count(MetricField):
```

Comment:
I added `Count`, `Avg`, etc. only as syntactic sugar. I believe for type safety it would be good enough to construct `MetricField("count", "foo")` (since `op` can only be of type `MetricOperationType`), but `Count("foo")` is IMO nicer to read.

Reply:
right, then I would write this:

```
class MetricField:
    @classmethod
    def count(cls, metric_name: str) -> MetricField: return cls("count", metric_name)
```

Reply:
Small thing: if you changed this to a field, you could collapse down your subclasses a bit.

```
@dataclass(frozen=True)
class MetricField:
    op: Optional[MetricOperationType] = dataclasses.field(init=False)
    metric_name: str

class Count(MetricField):
    op: Optional[MetricOperationType] = "count"


c = Count("some_metric")
# Count(op="count", metric_name="some_metric")
```

Depends on whether you go with the classmethod or not. Just thought I'd mention it.

Reply:
I would really check if we can make those classmethods, I don't think the experience is great having to import all of those types (and i think we don't typically import modules as such)

Reply:
Thanks @evanh! Is there a way I can use `dataclasses.field` while still allowing `MetricField` to be constructed generically with two arguments? I looked at `And` and `BooleanCondition` in snuba_sdk, but was discouraged by the fact that we need to define both fields in the subclass there:

https://github.com/getsentry/snuba-sdk/blob/be0a2e81ac524f25cb4514a6fd9c7abdd9fd3078/snuba_sdk/conditions.py#L133-L136

Reply:
@untitaker so the good thing is, you don't _have_ to import all those types, because you can always just write `MetricField("count", "foo")`. But I'm totally fine with removing the helper classes for now. I see little benefit in adding classmethods, because `MetricField.count("foo")` and `MetricField("count", "foo")` are equally readable in my opinion, and the type checker will catch typos in both cases.